lib/Controller/SocialApiController.php

```diff
@@ -0,0 +1,188 @@
+	 * @param string type which information to link to (avatar, ...)
+	 * @return string
+	 */
+	protected function getSocialConnector($socialentry, $type) : ?string {
+		if (!is_array($socialentry)) {
+			throw new Exception("socialentry format missmatch"); // FIXME: the Exceptions seem not to work as expected...
+		}
+
+		$candidate = $socialentry[3];
+		$network   = $socialentry[1]['type'];
+		$connector = null;
+
+		if (is_array($network)) { $network = $network[0]; }
+
+		// get profile-id
+		switch ($network) {
+			case "facebook":
+				$candidate = basename($candidate);
+				if (!ctype_digit($candidate)) {
+					// TODO: determine facebook profile id from username
+					throw new Exception("facebook profile-id expected to be a number, not %s", $candidate);
+				}
+				break;
+			default:
+				throw new Exception("%s not implemented", $network);
+		}
+
+		// build connector
+		switch ($network) {
+			case "facebook":
+				switch ($type) {
+					case "avatar":
+						$connector = "https://graph.facebook.com/" . ($candidate) . "/picture?width=720";
+						break;
+					default:
+						throw new Exception("%s for %s not implemented", $type, $network);
+				}
+				break;
+			default:
+				throw new Exception("Unexpected error building the connector for %s", $network);
+		}
+
+		return ($connector);
+
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 *
+	 * Retrieves social profile data for a contact
+	 *
+	 * @param string addressbookId identifier of the addressbook
+	 * @param string contact identifier of the contact
+	 * @param string type which information to get: avatar, ...
+	 */
+	public function fetch($addressbookId, $contactId, $type) {
+
+		$url = null;
+		$response = 404;
+
+		try {
+			// get social parameters from contact
+
+			/* TODO - port from javascript...
+			const jCal = this.contact.jCal
+			const socialentries = jCal[1].filter(props => props[0] === 'x-socialprofile')
+			socialentries.forEach(getSocialConnector)
+			*/
+			// FIXME static for testing:
+			$socialprofile = array("foo",array("type" => "facebook"),"bar","https://facebook.com/4");
+
+			try {
+				$url = $this->getSocialConnector($socialprofile, $type);
+			}
+			catch (Exception $e) {
+				$response = 500;
+				throw new Exception($e->getMessage());
+			}
+
+			if (empty($url)) {
+				$response = 500;
+				throw new Exception('not implemented');
+			}
+
+			$host = parse_url($url);
+			if (!$host) {
+				$response = 404;
+				throw new Exception('Could not parse URL');
+			}
+			$opts = [
+				"http" => [
+					"method" => "GET",
+					"header" => "User-Agent: Nextcloud Contacts App"
+				]
+			];
+			$context = stream_context_create($opts);
+			$socialdata = file_get_contents($url, false, $context);
+			if (!$socialdata) {
+				$response = 404;
+				throw new Exception('Could not parse URL');
+			}
+
+			$response = 200;
+
+			switch ($type) {
+				case "avatar":
+					header("Content-type:image/png");
+					break;
```

Comment:
I think It should be better if we just save the contacts directly too.
Sure we can return the image if we want, but since we might go for a background job after, it would make more sense to do all the logic here (meaning fetch photo & save vcard) instead of retrieving the image and downloading it on the frontend, then saving it again :)

Reply:
Now that I update everything in the backend, I need to refresh the contact for the user. Is there a routine for that by chance?

Reply:
Depending on the jsonResponse, (success or not), you ccan trigger a contact refresh with `updateContact`
https://github.com/nextcloud/contacts/blob/df9e046a18dd0871e2c18ea0bc8abb8ce0872c2f/src/components/ContactDetails.vue#L477

Reply:
And then trigger the full refresh: 
						await this.$store.dispatch('fetchFullContact', { contact })
https://github.com/nextcloud/contacts/blob/df9e046a18dd0871e2c18ea0bc8abb8ce0872c2f/src/components/ContactDetails.vue#L549

Make sure to update the contact afterwards in ContactsDetails :)

Reply:
Because you can trigger the fetchFullContact from the ContactDetailsAvatar component, but yu will still need to warn the ContactsDetail that the contact have changed, not sure it will work out of the box (try it, if just fetching the new update the photo, then it's fine! :) )

Reply:
That is what I tried, but `this.$store.getters.getContact(key)` seems to give me the local (outdated) copy and not the one from the server. Tried to call the refreshContact-function, but as I am new to vue I couldn't determine how to do so.

Reply:
Yes, thi sis what `this.$store.dispatch('fetchFullContact', { contact }) ` is for :)

Reply:
I tried to put all the pieces together, but it just doesn't refresh the view.

Reply:
I see you're still missing the `updateLocalContact` listener.
See my comments in https://github.com/nextcloud/contacts/pull/1580#discussion_r410679861

Reply:
I have it here locally, but didn't want to push it as it didn't change things.

Reply:
Does it create an error?
Have you made sure the updated contact object contains the new photo?

Reply:
There is no error. I think the problem is that I am not getting the updated Contact. The fetchFullContact seems to have no effect (I also tested by changing the FN-attribute)

Reply:
Ah, well, fetchFullContact doesn't return the updated contacthttps://github.com/nextcloud/contacts/blob/894ec80949ffe5845f72f3cb9a69f125fb02513f/src/store/contacts.js#L375-L385

Reply:
Sorry, I don't get it. Tried this (after the update in the background) to see what's happening
```
const key = contactId + '~' + addressbookId
let contact = this.$store.getters.getContact(key)
this.$store.dispatch('fetchFullContact', { contact })
contact = this.$store.getters.getContact(key)
this.$emit('updateLocalContact', contact)
```
but the contact stays constant the whole time.

Reply:
If you refreshes, does the changes appear?

Reply:
Yes, the contact is changed. A reload shows everything as expected. And if I modify the contact beforehand manually the server refuses the changes and proposes a refresh as it detects the background modification well.

Reply:
Can you push your changes so I can check?